tests/lib/config/flat-config-array.js

```diff
@@ -0,0 +1,1413 @@
+//-----------------------------------------------------------------------------
+
+const baseConfig = {
+    plugins: {
+        "@": {
+            rules: {
+                foo: {
+                    schema: {
+                        type: "array",
+                        items: [
+                            {
+                                enum: ["always", "never"]
+                            }
+                        ],
+                        minItems: 0,
+                        maxItems: 1
+                    }
+
+                },
+                bar: {
+
+                },
+                baz: {
+
+                },
+
+                // old-style
+                boom() {},
+
+                foo2: {
+                    schema: {
+                        type: "array",
+                        items: {
+                            type: "string"
+                        },
+                        uniqueItems: true,
+                        minItems: 1
+                    }
+                }
+            }
+        }
+    }
+};
+
+/**
+ * Creates a config array with the correct default options.
+ * @param {*[]} configs An array of configs to use in the config array.
+ * @returns {FlatConfigArray} The config array;
+ */
+function createFlatConfigArray(configs) {
+    return new FlatConfigArray(configs, {
+        basePath: __dirname,
+        baseConfig
+    });
+}
+
+/**
+ * Asserts that a given set of configs will be merged into the given
+ * result config.
+ * @param {*[]} values An array of configs to use in the config array.
+ * @param {Object} result The expected merged result of the configs.
+ * @returns {void}
+ * @throws {AssertionError} If the actual result doesn't match the
+ *      expected result.
+ */
+async function assertMergedResult(values, result) {
+    const configs = createFlatConfigArray(values);
+
+    await configs.normalize();
+
+    const config = configs.getConfig("foo.js");
+
+    assert.deepStrictEqual(config, result);
+}
+
+/**
+ * Asserts that a given set of configs results in an invalid config.
+ * @param {*[]} values An array of configs to use in the config array.
+ * @param {string|RegExp} message The expected error message.
+ * @returns {void}
+ * @throws {AssertionError} If the config is valid or if the error
+ *      has an unexpected message.
+ */
+async function assertInvalidConfig(values, message) {
+    const configs = createFlatConfigArray(values);
+
+    await configs.normalize();
+
+    assert.throws(() => {
+        configs.getConfig("foo.js");
+    }, message);
+}
+
+/**
+ * Normalizes the rule configs to an array with severity to match
+ * how Flat Config merges rule options.
+ * @param {Object} rulesConfig The rules config portion of a config.
+ * @returns {Array} The rules config object.
+ */
+function normalizeRuleConfig(rulesConfig) {
+    for (const ruleId of Object.keys(rulesConfig)) {
+        rulesConfig[ruleId] = [2];
+    }
+
+    return rulesConfig;
+}
```

Comment:
Some tests unrelated to this PR [are failing](https://github.com/eslint/eslint/pull/14321/checks?check_run_id=2494731957), probably because this directly modifies the given object (`eslint-recommended` and `eslint-all`).

Reply:
Doh! Yup, I'll fix that. Thanks.